include/zephyr/logging/log.h

```diff
@@ -89,13 +90,189 @@ extern "C" {
  */
 #define LOG_WRN_ONCE(...)					\
 	do {							\
-		static uint8_t __warned;			\
-		if (unlikely(__warned == 0)) {			\
+		static atomic_t __warned = ATOMIC_INIT(0);	\
+		if (unlikely(atomic_cas(&__warned, 0, 1))) {	\
 			Z_LOG(LOG_LEVEL_WRN, __VA_ARGS__);	\
-			__warned = 1;				\
 		}						\
 	} while (0)
 
+#if defined(CONFIG_LOG_RATELIMIT_ENABLE)
+/**
+ * @brief Writes a WARNING level message to the log with rate limiting.
+ *
+ * @details It's meant for situations that warrant investigation but could clutter
+ * the logs if output too frequently. The message will be logged at most once
+ * per specified interval in milliseconds.
+ *
+ * @param _rate_ms Minimum interval in milliseconds between log messages.
+ * @param ... A string optionally containing printk valid conversion specifier,
+ * followed by as many values as specifiers.
+ */
+#define LOG_WRN_RATELIMITED_RATE(_rate_ms, ...)					\
+	do {									\
+		static atomic_t __last_log_time = ATOMIC_INIT(0);		\
+		uint32_t __now = k_uptime_get_32();				\
+		uint32_t __last = atomic_get(&__last_log_time);		\
+		if (unlikely(__now - __last >= (_rate_ms))) {		\
+			if (atomic_cas(&__last_log_time, __last, __now)) {	\
+				Z_LOG(LOG_LEVEL_WRN, __VA_ARGS__);		\
```

Comment:
Not sure if this is possible or even desired, but in my head it would be great if the message output the number of skipped prints since the previous output.
```
Z_LOG(LOG_LEVEL_WRN, "Skipped %d: " GET_ARG_N(0, __VA_ARGS__), __skipped, GET_ARGS_LESS_N(1, __VA_ARGS__));
```